x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtValidateUtil.java

```diff
@@ -280,22 +283,41 @@ static void validateExpiredTime(final Date exp, final Date now, final long allow
      * @throws Exception Error if JWKs fail to validate the Signed JWT.
      */
     public static void validateSignature(final SignedJWT jwt, final List<JWK> jwks) throws Exception {
+        assert jwks != null && jwks.isEmpty() == false : "Caller must provide a non-empty JWK list";
         final String id = jwt.getHeader().getKeyID();
         final JWSAlgorithm alg = jwt.getHeader().getAlgorithm();
         LOGGER.trace("JWKs [" + jwks.size() + "], JWT KID [ + id + ], and JWT Algorithm [" + alg.getName() + "] before filters.");
 
+        // If JWT has optional kid header, and realm JWKs have optional kid attribute, any mismatches JWT.kid vs JWK.kid can be ignored.
+        // Keep any JWKs if JWK optional kid attribute is missing. Keep all JWKs if JWT optional kid header is missing.
         final List<JWK> jwksKid = jwks.stream().filter(j -> ((id == null) || (j.getKeyID() == null) || (id.equals(j.getKeyID())))).toList();
-        LOGGER.trace("JWKs [" + jwksKid.size() + "] after KID [" + id + "||null] filter.");
+        LOGGER.debug("JWKs [" + jwksKid.size() + "] after KID [" + id + "||null] filter.");
```

Comment:
I think these logs reallly should be trace level. They are step by step information. It does not fit [our guide](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md#logging) for debug logging. It's also to verbose since it is for every request that autenticates with JWT.

Reply:
Do we have good guidelines for log levels? I have received different opinions on log levels. I keep having to toggle log levels, because I can't find any good guidelines. There seems to be a lot of unwritten rules and assumptions people make, and I don't know what they are.

Here is the issue. During authenticate, if we filtered JWKs in the realm, I want the customer to diagnose that type of authc failure themselves. In other words, I want to avoid support issues coming in for diagnosing this type of authc error.

There is a tradeoff for picking DEBUG vs TRACE, but I don't know what that tradeoff is.

Reply:
https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md#log-levels

I agree this is a trace log.

Reply:
Resolved.